### DISCUSSION TITLE
Converter for support for all Javet objects

### DISCUSSION THREAD

**Discussion by Cormanz** [caoccao/Javet#61] [Ideas]
Hello, I'd like to propose a suggestion! I know Javet has support for [Interception](https://github.com/caoccao/Javet/blob/main/docs/tutorial/interception.rst), but coming from previously using GraalJS, being able to interop with **any java object** is just very, very convenient. Perhaps Javet could implement such a converter (I've attempted to but had some problems) which converts any Java object to a V8 object.

My idea of how this could work is like so:

You take some Java object, use Reflection to extract all of its methods, and you make the V8 object have those methods too. Then, in JS, when I call `javaObject.meth(1)`, it's going to get the actual Java object and call that method.

I think this would be really useful since it would allow for even more easily interoping between JS and Java.

**Comment by caoccao** [comment]
Thank you for raising this feature.

I intended to leave that to Javet users for 2 major reasons:

1. That brings considerable performance overhead. Currently JavetObjectConverter doesn't register property/function interceptors so that performance is excellent without runtime reflection.
2. Java allows overloading functions whereas JavaScript doesn't. Moreover, JavaScript allows passing less arguments. So, guessing which Java method is the one requires considerable coding effort. If it's not designed well, stack-overflow may occur.

As you've met, I think it's almost impossible to achieve that without support at JNI level. I'll consider doing some experiments. Of course, that takes some time.

**Reply by Cormanz** [reply]
If you're wondering, I don't at all suggest for this to be the default converter, but as an alternative one you could use with `.setConverter(new FullInteropJavaObjectConverter())` I just think it would be very useful.

1. This is _absolutely_ true, but for that performance cost, it can make life a lot easier too, especially for interoping large Java libraries and V8
2. This is true, but using some type checking with reflection on the arguments, it might be possible.

> As you've met, I think it's almost impossible to achieve that without support at JNI level. I'll consider doing some experiments. Of course, that takes some time.
Appreciated! For me, I actually have this working considerably well (to the point where I can do interop on Java classes), but the main issue I had a while back was for some reason, trying to use `new Thread(() => console.log('JS')).start()` would break.

I'll probably do my own experiments on this.

**Reply by caoccao** [reply]
You are right. It won't be the default offering as the performance overhead is considerable.

Regarding the threading API, I would recommend having a thread pool simulating it instead of exposing it directly. That's also the way Chrome and Node.js take, I suppose.

**Reply by caoccao** [reply]
Stay tuned. Native support is on the way.

**Reply by Cormanz** [reply]
That's _extremely_ exciting, since it's basically a necessity for my use case!

Thank you so much for working so hard on this project, I'm nearly 100% going to switch from GraalJS to this

**Reply by caoccao** [reply]
The idea is to have native [Proxy](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Proxy) set up universal interception. It may consist of the following 2 steps.

1. Provide full support to proxy via this [commit](https://github.com/caoccao/Javet/commit/8c1987656cbc9b78773e29d51d7a52019b21643f).
2. Provide a converter which is able to recursively convert Java objects from / to JavaScript objects. (TODO)

I'll be working on step 2 soon.

**Reply by Cormanz** [reply]
That's a different approach to what I tried, which would:

- Use reflection to find fields and methods
- Add those to a new V8ValueObject
- Whenever those are called, proxy those to the java method. If multiple methods exist, find the one which matches the types of the args you specified

Both approaches have benefits though, good luck on this!

**Reply by caoccao** [reply]
The `JavetProxyConverter` and related document has been included in this [commit](https://github.com/caoccao/Javet/commit/589e9887e2f03024c57151d8984866154a4214af).

Here are few issues in your approach for your reference.

1. It will face issues when dealing with overloaded methods. As JS doesn't have `overload`, Javet needs to guess which overloaded method is the most suitable one.
2. Its performance is not good as there are too many callback to be registered.
3. Its memory footprint is a little bit higher as every callback context has to be stored.

Please wait for the next release, and let me know if you have any questions.

**Reply by Cormanz** [reply]
That's true, the proxy converter is probably a better idea. Saw the new commit since I have this project subscribed/followed on github, thank you so much for adding this!

One more thing - perhaps for security purposes, the converter could be initialized in such a way that only specific packages and/or classes can be converted to/from Java?

**Reply by caoccao** [reply]
Yes, regarding the security, there are a few ways of achieving that. Subclassing, config, ... I'll find one or multiple suitable ways without performance overhead later.

**Reply by Cormanz** [reply]
Maybe one idea of what you could do is to allow for a lambda to be passed as an argument, which would take in the class instance, and return a boolean of if that lookup should be allowed or not.

**Reply by caoccao** [reply]
I decided to recommend subclassing for complete customization in response to the following challenges.

1. Class level filter customization
2. Generic template level filter customization
3. Method level filter customization
4. Value level filter customization

My current thought is it's quite hard to satisfy them in a simple way. So subclassing seems to be the only way, though it is not ideal.

Please refer to this [commit](https://github.com/caoccao/Javet/commit/f7e8a38d2891d22e6d1a8d052fa2fb9c81794f46) for detail.

**Comment by caoccao** [comment]
Now, it's included in v0.9.6. Enjoy.

**Reply by Cormanz** [reply]
Appreciated! Thank you!
